### ISSUE TITLE
Program of Courses option with Licences not working

### ISSUE THREAD

**Issue by lukebond68** [iomad/iomad#1319]
When you create a licence with a program of courses you are not able to select any users to give the licence to through "User Licence Allocations", and the number of licences available is incorrect and doesn't show as an integer eg 8.3333333 / 8.333333  licences can be allocated when it should show 10/10)

**Comment by turf212** [comment]
This is due to the fact that even though a program of courses is a nominal 1 to many of license to courses there is in actuality one license provided for each user for each course within the database and the number of "real" licenses are then calculated from it.  It sounds like only some of the licenses for a course have been provided to a user.  This is possibly because they already have another license for that course so it's skipped them.  Can you please check if that is the case as, if so, if it's a program of courses then the user shouldn't get displayed as a potential user if they already have a license for one of the program courses.

**Comment by fryeguy63** [comment]
I'm seeing the same issue. In 3.3 (I think) I updated a bit of code in the API function block_iomad_company_admin_create_licenses so that the math would be "correct" for my implementation. It works well in my current site. I just copied my system to a test server and updated to 3.8 to test. The problem is here again. Couldn't the code check if it is a program of courses and correct the math for the actual number to add? In my case it is always a program of courses so I can simply adjust the calculation within the API. If I remember, I simply multiplied the number of courses in the program of courses by the number of licenses. For me, it currently does not work in the API or within the UI.

**Comment by fryeguy63** [comment]
I went back into my production system and found the code in /blocks/iomad_company_admin/externallib.php inside the company_admin_create_license function:

```
public static function create_licenses($licenses = array()) {
        global $DB, $USER;

        $params = self::validate_parameters(self::create_licenses_parameters(), array('licenses' => $licenses));

        // Get/check context/capability
        $context = context_system::instance();
        self::validate_context($context);
        require_capability('block/iomad_company_admin:edit_licenses', $context);

        // Array to return newly created records
        $licenseinfo = array();

        foreach ($params['licenses'] as $license) {
                                                                                                                                                                                                
            // deal with the license quantity for program of courses                                                                                                                            
           if($license['program'] == TRUE) {
                $license['allocation'] = $license['allocation'] * count($license['courses']);
            }                                                                                                                                                                                                                                                                                                                                                                       // Create the License record                                                                                                                                                        $licenseid = $DB->insert_record('companylicense', $license);
```

By multiplying the number of licenses by the number of courses, the license quantity should work correctly.

Another issue is in the /blocks/iomad_company_admin/classes/forms/company_license_users_form.php -
https://github.com/iomad/iomad/blob/c8e9c951f73cffcfb5cbf75e6450cad7b86d49c9/blocks/iomad_company_admin/classes/forms/company_license_users_form.php#L196-L197
In my old code the formula for available licensese is **license->allocation - license->used / count(courseselect)** and the formula for total licenses is **license->allocation / count(courseselect)**.
In the 3.8 code each number has "1" subtracted. This makes no sense. The other issue is that $this->courseselect appears to only return visible courses even though courses that are not visible can be assigned to a license (for future use). So license allocation does not correspond with $this->courseselect and the math is wrong.

**Comment by turf212** [comment]
Thanks for the code for the WS calls.  I'd initially though that the calculation would have been done before the call was made but this also makes more sense (as to most people the actual number is never seen).

The lines where the calculation is different in 3.8 is because the license allocation page was reworked a few months back as allowing for license allocation (non program_ to any users and many courses was broken.  The list of courses included an "all" to there was one more entry in the course selector that just the list of courses.  I'll check that the calculation for program of courses is correct, it may be that this has had an inadvertent impact.  Probably have to check if the license is a program and update the modifier as -0 instead.

**Comment by fryeguy63** [comment]
Now that makes sense. I think you are correct that you need to check if it is a program of courses and the calc may be slightly different.

**Comment by turf212** [comment]
Interestingly, this is all working fine in 3.6 and 3.7 code. :/

**Comment by turf212** [comment]
OK.  So the issue which you are falling foul of is that there is now checking in the code to not assign a license to a user where they already have a license for that course which is not finished with.  SO, if you assign a program of courses to a user who has such a license assignation, the license for that course is skipped and the calculation gets broken.  So, there also needs to be a check so that users who this is impacting do not get listed in the potential users.

FYI where there isn't a licensing clash, the code works as expected.

**Comment by turf212** [comment]
I've added the fix for the externalib in #1331 1331

**Comment by lukebond68** [comment]
Thanks both for your help with this!

Kind regards
Luke Bond
